Sources/Spezi/Dependencies/Property/DependencyPropertyWrapper.swift

```diff
@@ -52,8 +52,35 @@ public class _DependencyPropertyWrapper<Value> { // swiftlint:disable:this type_
         self.dependencies = dependencies
     }
 
+    fileprivate convenience init<T>(_ context: DependencyContext<T>) {
+        self.init(DependencyCollection(context))
+    }
+
+    /// Create an optional dependency.
+    /// - Parameter dependencyType: The wrapped type of the optional dependency.
+    public convenience init<T>(_ dependencyType: T.Type) where Value == T?, T: Module {
+        self.init(DependencyCollection(DependencyContext(for: T.self, type: .optional)))
+    }
+
+    /// Create an optional dependency with a default value.
+    /// - Parameters:
+    ///   - dependencyType: The wrapped type of the optional dependency.
+    ///   - defaultValue: The default value that is used if no instance was supplied otherwise.
+    public convenience init<T>(
+        wrappedValue defaultValue: @escaping @autoclosure () -> T,
+        _ dependencyType: T.Type = T.self
+    ) where Value == T?, T: Module {
+        self.init(DependencyContext(for: T.self, type: .optional, defaultValue: defaultValue))
+    }
+
     /// Declare a dependency to a module that can provide a default value on its own.
-    public convenience init() where Value: Module & DefaultInitializable {
+    @available(
+        *, deprecated, renamed: "init(wrappedValue:_:)",
+         message: "Please explicitly specify the default value for 'DefaultInitializable' modules"
```

Comment:
Wondering what the reason is to explicitly provide warnings about this use case. A module with DefaultInitializable sort of manages the default value for your and allows you to worry less about this in this use case. What would be the reason why we want to avoid this behavior in the future?

Reply:
There are two driving factors behind this.

One is technical, as Swift has troubles to detect that a property wrapper is only initialized in the init and a default initializer shouldn't be used. The following code snippet, where `ModuleB` doesn't conform to `DefaultInitializable` currently doesn't compile:
```swift
class Module1: Module {
    @Dependency var moduleB: ModuleB // error: No exact matches in call to initializer

    init() {
        _moduleB = Dependency(wrappedValue: ModuleB(/* ... */))
    }
}
```
Event though ModuleB doesn't conform to `DefaultInitializable` nor is declared as an optional, it tries to apply these default inits and fails. If you remove the default inits, this compiles without any issues.

The second reason is more of a semantic one. The goal was to make default values just more explicit and not a "magic thing" to better differentiate them from required dependencies that do not specify a default value (which was introduced with this PR).

Interested what you think about this. Happy to talk about this in the meeting later 👍

Reply:
I see; thanks for the context @Supereg!

The Swift overload issue is a bit unfortunate.
At the same time, it seems a bit painful to always add a default value for all modules, especially for simple ones.

I am wondering if we can just incorporate this in the general initializer that doesn't take a wrapped value. If the module conforms to `DefaultInitializable`, it would not crash on startup with a precondition, but it would rather load the `DefaultInitializable` module? We still use the same syntax as the SwiftUI `@Environment(...)` that takes the type as an argument but could just implicitly load elements if the module defined is confirming to `DefaultInitializable` instead of a precondition failure?

We can still add an option to provide a default wrapped value if a user wants to do that; this should take precedence over the usage of a `DefaultInitializable` conformance.

A general thing that I am wondering: How do we do conflict resolution if one module has a default value with a parameter A and another one with B? As far as I remember my initial implementation and as far as I can see from the current changes, we would assume that whatever module is initialized first will be used across all parts of the modules, and therefore the order of initialization in the Configuration is important?

Reply:
> I am wondering if we can just incorporate this in the general initializer that doesn't take a wrapped value. If the module conforms to `DefaultInitializable`, it would not crash on startup with a precondition, but it would rather load the `DefaultInitializable` module? We still use the same syntax as the SwiftUI `@Environment(...)` that takes the type as an argument but could just implicitly load elements if the module defined is confirming to `DefaultInitializable` instead of a precondition failure?

Sounds good to me 👍 You are right, that `DefaultInitializable` conformance is not necessary to know at compile time.

> A general thing that I am wondering: How do we do conflict resolution if one module has a default value with a parameter A and another one with B? As far as I remember my initial implementation and as far as I can see from the current changes, we would assume that whatever module is initialized first will be used across all parts of the modules, and therefore the order of initialization in the Configuration is important?

Yes, it will use the first value. The only change here, is that due to using breadth-first search instead of depth-first search this yields way more predictable results or rather what one would intuitively expect. So it uses the "first" and "outermost" default value were previously it could have been that it would use the default value of some deeply nested `@Dependency` declaration from the first module.

Reply:
Thanks for adopting it to use `DefaultInitializable` when possible within the other initializer.
We should then probably adjust the error message here to point to the `@Dependency(...) var ...` initializers to point people to slowly upgrade to that syntax.

Makes sense, that is more intuitive 👍

Reply:
Updated the message and PR description to reflect these changes 👍

Reply:
Thank you @Supereg!